src/libraries/System.Net.Ping/src/System/Net/NetworkInformation/Ping.cs

```diff
@@ -408,16 +468,28 @@ private PingReply GetAddressAndSend(string hostNameOrAddress, int timeout, byte[
             }
         }
 
-        private async Task<PingReply> GetAddressAndSendAsync(string hostNameOrAddress, int timeout, byte[] buffer, PingOptions? options)
+        private async Task<PingReply> SendPingAsyncInternal<TArg>(
+            TArg getAddressArg,
+            Func<TArg, CancellationToken, ValueTask<IPAddress>> getAddress,
+            int timeout,
+            byte[] buffer,
+            PingOptions? options,
+            CancellationToken cancellationToken)
         {
+            cancellationToken.ThrowIfCancellationRequested();
+
             CheckStart();
             try
             {
-                IPAddress[] addresses = await Dns.GetHostAddressesAsync(hostNameOrAddress).ConfigureAwait(false);
-                Task<PingReply> pingReplyTask = SendPingAsyncCore(addresses[0], buffer, timeout, options);
-                return await pingReplyTask.ConfigureAwait(false);
+                using CancellationTokenRegistration _ = cancellationToken.Register(static state => ((Ping)state!).SetCanceled(), this);
+
+                IPAddress address = await getAddress(getAddressArg, _timeoutOrCancellationSource!.Token).ConfigureAwait(false);
+
+                Task<PingReply> pingTask = SendPingAsyncCore(address, buffer, timeout, options);
+                _timeoutOrCancellationSource.CancelAfter(timeout);
```

Comment:
Doing this here means the timeout won't apply to the GetHostAddressesAsync inside getAddress.  Why is that ok / desirable?  I'm also unclear as to why this is being done like this after the SendPingAsyncCore.  Normally this kind of ordering would be to avoid a race condition, but that doesn't apply here.

Reply:
Both of these are attempts to maintain the existing behavior / be true to the documentation:

* The old code does not apply the timeout to `Dns.GetHostAddressesAsync` (see https://source.dot.net/#System.Net.Ping/System/Net/NetworkInformation/Ping.cs,646).
* The documentation for [the `timeout` parameter](https://docs.microsoft.com/en-us/dotnet/api/system.net.networkinformation.ping.sendpingasync?view=net-6.0#system-net-networkinformation-ping-sendpingasync(system-string-system-int32-system-byte()-system-net-networkinformation-pingoptions)) says: "The maximum number of milliseconds (*after sending the echo message*) to wait for the ICMP echo reply message". 

I feel like the timeout is supposed to be inherent to the ping request, rather than being a general timeout for the entire operation. In contrast, I think the `CancellationToken` is the way for the caller to abort the operation whenever they want. Thoughts?

Reply:
I'll defer to @dotnet/ncl on the desired behavior here.  I expect it was documented that way because there was literally no way before for the Dns operations to be timed out.  But there is now.  The question then is whether we want to update the timeout to apply to the whole operation or just to the waiting for the ICMP echo reply message; I'm ok either way.

I've not looked; are we consistent with the timeout only applying to the ICMP echo reply across all implementations?

Reply:
Personally, I feel like it is better to keep the timeout only for the ICMP echo reply. As noted, users can use the `CancellationToken` to cancel the operation *x* ms from the start of it, and if we make the `timeout` parameter cover the DNS resolution as well, users may get different results depending on whether DNS was fast enough or was/wasn't cached.